src/python/pants/backend/codegen/thrift/lib/apache_thrift_gen_base.py

```diff
@@ -12,112 +12,67 @@
 
 from twitter.common.collections import OrderedSet
 
-from pants.backend.codegen.thrift.java.java_thrift_library import JavaThriftLibrary
-from pants.backend.codegen.thrift.java.thrift_defaults import ThriftDefaults
-from pants.backend.jvm.targets.java_library import JavaLibrary
 from pants.base.build_environment import get_buildroot
-from pants.base.exceptions import TargetDefinitionException, TaskError
+from pants.base.exceptions import TaskError
 from pants.base.workunit import WorkUnitLabel
 from pants.binaries.thrift_binary import ThriftBinary
+from pants.option.custom_types import target_option
 from pants.task.simple_codegen_task import SimpleCodegenTask
 from pants.util.memo import memoized_property
 
 
-class ApacheThriftGen(SimpleCodegenTask):
+class ApacheThriftGenBase(SimpleCodegenTask):
 
-  _COMPILER = 'thrift'
-  _LANG = 'java'
-  _RPC_STYLE = 'sync'
+  # The type of target this codegen task acts on.
+  # E.g., JavaThriftLibrary, PythonThriftLibrary.
+  thrift_library_target_type = None
+
+  # The name of the thrift generator to use. Subclasses must set.
+  # E.g., java, py (see `thrift -help` for all available generators).
+  thrift_generator = None
+
+  # Subclasses may set their own default generator options.
+  default_gen_options_map = None
 
   @classmethod
   def register_options(cls, register):
-    super(ApacheThriftGen, cls).register_options(register)
+    super(ApacheThriftGenBase, cls).register_options(register)
 
     # NB: As of thrift 0.9.2 there is 1 warning that -strict promotes to an error - missing a
     # struct field id.  If an artifact was cached with strict off, we must re-gen with strict on
     # since this case may be present and need to generate a thrift compile error.
     register('--strict', default=True, fingerprint=True, type=bool,
              help='Run thrift compiler with strict warnings.')
-
+    # The old --gen-options was string-typed, so we keep it that way for backwards compatibility,
+    # and reluctantly use the clunky name --gen-options-map for the new, map-typed options.
+    # TODO: Once --gen-options is gone, do a deprecation cycle to restore the old name.
     register('--gen-options', advanced=True, fingerprint=True,
-             help='Use these apache thrift java gen options.')
-    register('--deps', advanced=True, type=list,
-             help='A list of specs pointing to dependencies of thrift generated java code.')
-    register('--service-deps', advanced=True, type=list,
-             help='A list of specs pointing to dependencies of thrift generated java service '
+             removal_version='1.5.0', removal_hint='Use --gen-options-map instead',
```

Comment:
For this deprecation and others, please see https://groups.google.com/d/msg/pants-committers/OrLkQOsUSDo/KYlpcrIaBAAJ ... sorry that I haven't published that anywhere yet.

In this case, I think that means targeting `1.5.0.dev0`.